module/signature/aggregation.go

```diff
@@ -3,21 +3,317 @@
 package signature
 
 import (
+	"errors"
 	"fmt"
+	"sync"
 
 	"github.com/onflow/flow-go/crypto"
 	"github.com/onflow/flow-go/crypto/hash"
+	"github.com/onflow/flow-go/engine"
 	"github.com/onflow/flow-go/module"
 )
 
+// SignatureAggregatorSameMessage aggregates BLS signatures of the same message from different signers.
+// The public keys and message are aggreed upon upfront.
+//
+// Currently, the module does not support signatures with multiplicity higher than 1. Each signer is allowed
+// to sign at most once.
+//
+// Aggregation uses BLS scheme. Mitigation against rogue attacks is done using Proof Of Possession (PoP)
+// This module does not verify PoPs of input public keys, it assumes verification was done outside this module.
+//
+// Implementation of SignatureAggregator is not thread-safe, the caller should
+// make sure the calls are concurrent safe.
+type SignatureAggregatorSameMessage struct {
+	message          []byte
+	hasher           hash.Hasher
+	n                int                // number of participants indexed from 0 to n-1
+	publicKeys       []crypto.PublicKey // keys indexed from 0 to n-1, signer i is assigned to public key i
+	indexToSignature map[int]string     // signatures indexed by the signer index
+}
+
+// NewSignatureAggregatorSameMessage returns a new SignatureAggregatorSameMessage structure.
+//
+// A new SignatureAggregatorSameMessage is needed for each set of public keys. If the key set changes,
+// a new structure needs to be instantiated. Participants are defined by their public keys, and are
+// indexed from 0 to n-1 where n is the length of the public key slice.
+// The function errors with engine.InvalidInputError if any input is invalid.
+func NewSignatureAggregatorSameMessage(
```

Comment:
Can we document sentinel errors and return values in general?

Reply:
Can you please clarify how I should document the errors ? I'm happy to update the doc.

Reply:
Exactly the same way you did in other places. Just write a godoc what errors are expected, unexpected and when expected errors can be seen.